Improve Flask integration

c-w opened the pull request:
Unlike the Django integration for Application Insights, the Flask integration currently is quite low-level: the developer has to add a custom WSGI middleware and enable trace and exception logging themselves.

A more common pattern for adding groups of new capabilities to Flask is using custom extensions that initialize themselves against the Flask application: http://flask.pocoo.org/docs/0.12/extensiondev/

As such, this commit adds a custom extension for Flask that enables a developer to very easily add Application Insights integration to their Flask application just by doing `AppInsights(app)`.

The extension enables:
- Request logging
- Trace logging
- Exception logging

SergeyKanzhelev commented on README.md:
```diff
@@ -169,20 +169,29 @@ enable('<YOUR INSTRUMENTATION KEY GOES HERE>')
+# instantiate the Flask application
 app = Flask(__name__)
-app.wsgi_app = WSGIApplication('<YOUR INSTRUMENTATION KEY GOES HERE>', app.wsgi_app)
+app.config['APPLICATION_INSIGHTS_KEY'] = '<YOUR INSTRUMENTATION KEY GOES HERE>'
```
Flask doesn't translate environment variables into the config automatically. So this comment is not very critical. On some environments in Azure like Web Apps instrumentation key will be set in environment variable `APPINSIGHTS_INSTRUMENTATIONKEY`. Name you use for `config` is different. Would it be more confusing to align with the name of environment variable or beneficial?

c-w replied:
Might have slightly misunderstood the original intent here so I also added the option to auto-load the value from the OS environment in 157a5b7.

SergeyKanzhelev commented on README.md:
```diff
@@ -169,20 +169,29 @@ enable('<YOUR INSTRUMENTATION KEY GOES HERE>')
+app.config['APPLICATION_INSIGHTS_KEY'] = '<YOUR INSTRUMENTATION KEY GOES HERE>'
+
+# log requests, traces and exceptions to the Application Insights service
+appinsights = AppInsights(app)
```
It will also be good to configure EndpointUrl here. Endpoint URL is used for reverse proxy and compliance scenarios.

c-w replied:
Good to know, done in aa0d8da.

OsvaldoRosado commented on README.md:
````diff
@@ -169,20 +169,29 @@ enable('<YOUR INSTRUMENTATION KEY GOES HERE>')
 ```python
 from flask import Flask
-from applicationinsights.requests import WSGIApplication
+from applicationinsights.flask.ext import AppInsights
````
Is the WSGI wrapper useful for other frameworks? We might not want to remove it from the docs.

c-w replied:
Added a link to the middleware in d0b2961. Also updated the middleware's docstring to not use Flask as the code example to avoid potential confusion.

OsvaldoRosado commented on README.md:
```diff
@@ -169,20 +169,29 @@ enable('<YOUR INSTRUMENTATION KEY GOES HERE>')
+# ensure that the telemetry gets sent
+@app.after_request
+def send_telemetry(response):
+    appinsights.flush()
```
Is this a good practice to recommend? Won't this impede batching?

c-w replied:
Removed in d0b2961.

c-w commented:
Hi @OsvaldoRosado and @SergeyKanzhelev. Thanks for the reviews. I should have addressed all the comments now. Do you have any other questions or concerns before this can be merged?

SergeyKanzhelev commented:
@c-w thank you!

c-w commented:
Thanks for the merge, @SergeyKanzhelev! Any ideas when you'll be able to do a new release to PyPI? I have a partner who's already using this new Flask integration for Application Insights and it would be great to get them onto an official dependency instead of a Github zip archive.

SergeyKanzhelev commented:
@c-w I'll try to release it tonight.

c-w commented:
Awesome! Thanks a lot!

SergeyKanzhelev commented:
@c-w check it out. Just published

c-w commented:
Great, thanks!
